pkg/gc/gc_state_manager.go

```diff
@@ -0,0 +1,567 @@
+	if keyspaceID == constant.NullKeyspaceID && barrierID == keypath.GCWorkerServiceSafePointID {
+		return nil, errs.ErrReservedGCBarrierID.GenWithStackByArgs(barrierID)
+	}
+	// Disallow empty barrierID
+	if len(barrierID) == 0 {
+		return nil, errs.ErrInvalidArgument.GenWithStackByArgs("barrierID", barrierID)
+	}
+
+	var deletedBarrier *endpoint.GCBarrier
+	err := m.gcMetaStorage.RunInGCStateTransaction(func(wb *endpoint.GCStateWriteBatch) error {
+		var err1 error
+		deletedBarrier, err1 = m.gcMetaStorage.LoadGCBarrier(keyspaceID, barrierID)
+		if err1 != nil {
+			return err1
+		}
+		return wb.DeleteGCBarrier(keyspaceID, barrierID)
+	})
+	return deletedBarrier, err
+}
+
+// getGCStateInTransaction gets all properties in GC states within a context of gcMetaStorage.RunInGCStateTransaction.
+// This read only and won't write anything to the GCStateWriteBatch. It still receives a write batch to ensure
+// it's running in a transactional context.
+func (m *GCStateManager) getGCStateInTransaction(keyspaceID uint32, _ *endpoint.GCStateWriteBatch) (GCState, error) {
+	result := GCState{
+		KeyspaceID: keyspaceID,
+	}
+	if keyspaceID != constant.NullKeyspaceID {
+		result.IsKeyspaceLevel = true
+	}
+
+	var err error
+	result.TxnSafePoint, err = m.gcMetaStorage.LoadTxnSafePoint(keyspaceID)
+	if err != nil {
+		return GCState{}, err
+	}
+
+	result.GCSafePoint, err = m.gcMetaStorage.LoadGCSafePoint(keyspaceID)
+	if err != nil {
+		return GCState{}, err
+	}
+
+	result.GCBarriers, err = m.gcMetaStorage.LoadAllGCBarriers(keyspaceID)
+	if err != nil {
+		return GCState{}, err
+	}
+
+	// For NullKeyspace, remove GC barrier whose barrierID is "gc_worker", which is only exists for providing
+	// compatibility with the old versions.
+	if keyspaceID == constant.NullKeyspaceID {
+		result.GCBarriers = slices.DeleteFunc(result.GCBarriers, func(b *endpoint.GCBarrier) bool {
+			return b.BarrierID == keypath.GCWorkerServiceSafePointID
+		})
+	}
+
+	return result, nil
+}
+
+// GetGCState returns the GC state of the given keyspace.
+//
+// When this method is called on a keyspace without keyspace-level GC enabled, it will be equivalent to calling it on
+// the NullKeyspace.
+func (m *GCStateManager) GetGCState(keyspaceID uint32) (GCState, error) {
+	keyspaceID, err := m.redirectKeyspace(keyspaceID, true)
+	if err != nil {
+		return GCState{}, err
+	}
+
+	var result GCState
+	err = m.gcMetaStorage.RunInGCStateTransaction(func(wb *endpoint.GCStateWriteBatch) error {
+		var err1 error
+		result, err1 = m.getGCStateInTransaction(keyspaceID, wb)
+		return err1
+	})
+
+	if err != nil {
+		return GCState{}, err
+	}
+
+	return result, nil
+}
+
+// saturatingDuration returns a duration calculated by multiplying the given `ratio` and `base`, truncated within the
+// range [0, math.MaxInt64] to avoid negative value and overflowing.
+func saturatingDuration(ratio int64, base time.Duration) time.Duration {
```

Comment:
You don't use this function in this pr. How about removing it now until you use it?

Reply:
Oh..
I will move it to the next PR where it's used